### ISSUE TITLE
Speed issue with IE11 - suggested fix

### ISSUE THREAD

**Issue by KevDevMan** [jhildenbiddle/css-vars-ponyfill#44]
Hi John,

When running on IE11 (11.492.16299.0) testers started seeing a "long running script" error on one of our Angular apps. On investigation, I found that Ponyfill was taking 60+ seconds to run on this app. After further investigation, I found the Ponyfill was taking seconds (from 10 to 60) on all our apps (in most cases unnoticed). All our apps use the same Angular core where almost all of our css is generated. 

Reason:
Angular likes to group css where the properties are the same. In my case, I had over 1900 selectors for "margin-left:-2px;", for another app it was 1200. Ponyfill was struggling to cope with this amount of data and there was a noticeable lag when it hit these lines.

My solution: 
I am not suggesting this is the best approach. But the speed increase was very noticeable (from 60 to under 1 second) 

I pushed the "onlyVars" setting to the "rule" function. If this is true, then I also check the "declarations" to see if they have a 'var(--XX' defined. If yes, I go on as normal. Other wise i return an empty node object.

below is my added function which is called at the start of the rule function
```
function hasVars() {
        // Get declaration block in line
        // same as match but does not remove from cssText
        var nextDeclarations = matchPreserve(/{([^}]*)}/);
       
        // if we have a var then allow to continue
        if (nextDeclarations != undefined && nextDeclarations.indexOf('var(--') !== -1) {
            return true;
        } else {
            // else remove the string from the css file and return false to go to next line
            match(/([^}]*)}/);
            return false;
        }
    }
```

**Comment by jhildenbiddle** [comment]
Hi @KevDevMan,

Thanks for taking the time to share the issue and your solution. Very much appreciated.

The solution above is a good way to improve performance when [`options.onlyVars`](https://github.com/jhildenbiddle/css-vars-ponyfill#optionsonlyvars) set to `true`. The issue is that this option is `false` by default to guarantee that the original cascade is maintained. From the README:

> When true, rulesets and declarations without a custom property value will be removed from the generated CSS, reducing CSS output size. This can significantly reduce the amount of CSS processed and output by the ponyfill, but runs the risk of breaking the original cascade order once the transformed values are appended to the DOM

Depending on how your CSS is structured, using `onlyVars:true` may or may not cause cascade issues. Here's an example CSS that would break:

```css
/* Original stylesheet */
:root {
  --color: red;
}

p {
  color: var(--color);
}

@media screen and (min-width: 1px) {
  p {
    color: green;
  }
}
```

With this stylesheet, the color of `<p>` text will be green. This is because of the `@media` `p` rule which has the same specificity as but comes *after* the `p { color: var(--color); }` rule.

When this above CSS is processed by the ponyfill with `onlyVars:true`, the following CSS will be appended to the DOM:

```css
/* Stylesheet appended by ponyfill with `onlyVars:true` */
p {
  color: red
}
```

All `<p>` text will be red, which is incorrect. This is because the ponyfill-generated CSS comes *after* and therefore overrides the previous `@media` `p` rule.  

If the ponyfill had been called with `onlyVars:false`, the following CSS would be appended to the DOM instead:

```css
/* Stylesheet appended by ponyfill with `onlyVars:false` */
p {
  color: red;
}

@media screen and (min-width: 1px) {
  p {
    color: green;
  }
}
```

Now the cascade is maintained because the `@media` rule is included with the ponyfill-generated CSS.

All that being said, it certainly won't hurt to make `onlyVars:true` scenario faster even if it isn't the "safe" way of calling the ponyfill. I can make this change fairly quickly so that you don't have to run with a hacked version of the ponyfill.

Long-term, I'd like to explore ways of improving the ponyfill's performance without having to resort to `onlyVars:true`. Specifically, I'm thinking that the entire parse/transform process could be greatly simplified by parsing the CSS for the following strings:

* `var(...)` for transformation
* `calc(...)` for [`options.fixNestedCalc`](https://github.com/jhildenbiddle/css-vars-ponyfill#optionsfixnestedcalc)
* `url(...)` for [`options.updateURLs`](https://github.com/jhildenbiddle/css-vars-ponyfill#optionsupdateurls)

Not sure when I'll be able to dive into that though. Hopefully before EOY.

**Comment by KevDevMan** [comment]
Hi @jhildenbiddle,

Thanks for getting back to me. Also i neglected to mention earlier - Thank you for all your work on this.

It would be great if you could include this in the next release. I would really appreciate it.

I understand the issue you have made. This does not effect us BUT it could effect others. My Solution is not 100% as it includes a selector list and list of **ALL** declarations regardless of value. 

You could alter my suggestion so that it ONLY adds declarations with var(..), calc(..) or (url...) to the declarations array. NOW, you can remove the onlyVars from the output section. Using the current logic, you will still get the @media.

Kevin

**Comment by KevDevMan** [comment]
Hi John,

Thanks for all your efforts here. I really appreciate it.

I am getting an error i was not getting before. I will try to explain.
the css throwing the error is. 

`@-ms-viewport {  width: device-width; }`

Due to the @ i am being sent to the at_rule() function as expected. In the old code, and the new code, this @ does not fall into any of the "at" categories so the "ret" is undefined. But in the new code, we are attempting to use this undefined code. (This worked for me in the old code as none of my "@" use css vars)

`if (ret.declarations) {`

so i am getting a "declarations is undefined" error.

Kind Regards

Kevin

**Comment by jhildenbiddle** [comment]
Hey @KevDevMan,

Should be fixed in 1.16.1. Sorry about that.
